[DNM] samples: nrf9160: lwm2m_client: sample demonstrating LwM2M

mike-scott opened the pull request:
This sample demonstrates the use of Zephyr's LwM2M client configured
for the nRF9160 DevKit to support: Buttons, Switches, Buzzer,
Connectivity Data and more.

To support FOTA capability the FIRMWARE object is implemented to
allow CoAP transmission of a signed combined binary of the secure
and non-secure samples.   MCUboot checks the signature on the next
boot and then moves that binary from slot1 to slot0 where it will
be booted.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
This PR requires these upstream changes in Zephyr: https://github.com/zephyrproject-rtos/zephyr/pull/17710

For now we can discuss review items here.

mike-scott commented:
As well as these PRs for nrfconnect:
https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/992
https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1005

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:572: WARNING:STATIC_CONST_CHAR_ARRAY: static char array declaration should probably be static const char
#572: FILE: samples/nrf9160/lwm2m_client/src/config.h:1:
+static char client_psk[] = "000102030405060708090a0b0c0d0e0f";

-:985: ERROR:SWITCH_CASE_INDENT_LEVEL: switch and case should be at the same indent
#985: FILE: samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_button.c:44:
+	switch (evt->button) {
+		case UI_BUTTON_1:
[...]
+		case UI_BUTTON_2:
[...]
+		case UI_SWITCH_1:
[...]
+		case UI_SWITCH_2:

-:2314: WARNING:NEW_TYPEDEFS: do not add new typedefs
#2314: FILE: samples/nrf9160/lwm2m_client/src/settings.h:15:
+typedef enum {

-:3143: WARNING:LONG_LINE: line over 80 characters
#3143: FILE: samples/nrf9160/lwm2m_client/src/ui/led_pwm.c:58:
+	[UI_LED_ERROR_BSD_IRREC] = LED_EFFECT_LED_BREATHE(UI_LED_ON_PERIOD_ERROR,

-:3205: WARNING:LONG_LINE: line over 80 characters
#3205: FILE: samples/nrf9160/lwm2m_client/src/ui/led_pwm.c:120:
+			__ASSERT_NO_MSG(leds.effect->steps[leds.effect_step].substep_count > 0);

- total: 1 errors, 4 warnings, 3253 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.



```

mike-scott commented:
Regarding checkpatch issue #1: I'm using the PSK with this function:
```
int nrf_inbuilt_key_write(nrf_sec_tag_t            sec_tag,
                          nrf_key_mgnt_cred_type_t cred_type,
                          uint8_t                * p_buffer,
                          uint16_t                 buffer_len);
```

If I add const I see this warning:
```
nrf/samples/nrf9160/lwm2m_client/src/main.c:240:10: warning: passing argument 3 of 'nrf_inbuilt_key_write' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
          client_psk, strlen(client_psk));
          ^~~~~~~~~~
In file included from nrf/samples/nrf9160/lwm2m_client/src/main.c:25:
nrfxlib/bsdlib/include/nrf_inbuilt_key.h:47:52: note: expected 'uint8_t *' {aka 'unsigned char *'} but argument is of type 'const char *'
                           uint8_t                * p_buffer,
                           ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
```

mike-scott commented:
@rlubos I realize that the PWM path under UI is completely untested.  Should I remove it for now?  Or leave it in and then clean it up once we get real HW testing?

mike-scott commented:
@mbolivar adding for review

rlubos commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_location.c:
```diff
@@ -0,0 +1,108 @@
+
+static void update_location_data(struct gps_data *nmea)
+{
+	LOG_DBG("nmea_data");
```
@joakimtoe Do you know if GPS socket provides us with NMEA string, or a structure with separate fields for latitude, longitude etc.? LWM2M expects to have these values in separate fields instead of being encoded in the NMEA string.

mike-scott replied:
@rlubos @joakimtoe From the GPS sample, it looks like if the modem is in GPS mode (XSYSTEMMODE=0,0,1,0) then I can use `nrf_recv` to fill a `nrf_gnss_data_frame_t structure`.  This structure contains `nrf_gnss_pvt_data_frame_t` which has longitude, latitude, etc.  That would be the best data format for the Location object in the LwM2M client.

I have to wonder if both LTE and GPS can be used at the same time?  There is no sample which shows that currently.

rlubos replied:
@mike-scott There's a limitation on GPS use - the modem needs to be in power saving mode in order to get measurements. From the nRF9160 Product Specification:
```
The GPS receiver supports GPS L1 C/A reception. Operation is time multiplexed with LTE 
modem, and it is possible to obtain the GPS position either while the LTE is in RRC Idle
mode or power saving mode (PSM), or when the LTE modem is completely deactivated.
```
https://infocenter.nordicsemi.com/topic/ps_nrf9160/gps.html

rlubos commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_security.c:
```diff
@@ -0,0 +1,68 @@
+#include <net/tls_credentials.h>
+#include "nrf_inbuilt_key.h"
+
+#define TLS_TAG			16842753
```
We should use a different tag here. This one is associated with a factory-assigned certificate/private key. I found the `asset_tracker` unable to connect after running lwm2m sample because it had PSK/identity assigned with the same tag - had to remove DTLS credentials manually.

mike-scott replied:
I reversed the tag to 35724861

rlubos commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_temp.c:
```diff
@@ -0,0 +1,97 @@
+static int read_temperature(struct device *temp_dev,
+			    struct float32_value *float_val)
+{
+	__unused const char *name = temp_dev->config->name;
```
What's `__unused`?

mike-scott replied:
Woops.  Removing __unused.  It's clearly used :P

mike-scott commented:
@rlubos thank you for the the detailed review!  I'll address these comments today.

mike-scott commented on samples/nrf9160/lwm2m_client/Kconfig:
```diff
@@ -0,0 +1,198 @@
+	  The minimum time in seconds that notifications can be sent to
+	  observers of the Connectivity Monitoring data.
+
+config APP_HOLD_TIME_GSP
```
nice catch.  fixed.

mike-scott commented on samples/nrf9160/lwm2m_client/Kconfig:
```diff
@@ -0,0 +1,198 @@
+module = APP
+module-dep = LOG
+module-str = Log level for sample
+source "subsys/logging/Kconfig.template.log_config"
```
I notice this is not consistent across the nrf tree.  But I will add ZEPHYR_BASE here so that at least it's consistent in the sample.

mike-scott commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_accelerometer.c:
```diff
@@ -0,0 +1,244 @@
+
+/**@brief Struct containing current orientation and 3 axis acceleration data. */
+struct orientation_detector_sensor_data {
+	double x;			    /**< X-axis acceleration [m/s^2]. */
```
yep.  will fix.

mike-scott commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_accelerometer.c:
```diff
@@ -0,0 +1,244 @@
+	for (i = 0; i < MEASUREMENT_ITERATIONS; i++) {
+		err = sensor_sample_fetch_chan(accel_dev, SENSOR_CHAN_ACCEL_Z);
+		if (err) {
+			printk("sensor_sample_fetch failed\n");
```
Thanks for catching.  I was changing these throughout where I used code from the asset_tracker.

mike-scott commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_accelerometer.c:
```diff
@@ -0,0 +1,244 @@
+				SENSOR_CHAN_ACCEL_Z, &accel_data[2]);
+
+		if (err) {
+			printk("sensor_channel_get failed\n");
```
Same

mike-scott commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_accelerometer.c:
```diff
@@ -0,0 +1,244 @@
+
+		err = sensor_channel_get(accel_dev,
+				SENSOR_CHAN_ACCEL_Z, &accel_data[2]);
+
```
fixed

mike-scott commented on samples/nrf9160/lwm2m_client/src/lwm2m/lwm2m_accelerometer.c:
```diff
@@ -0,0 +1,244 @@
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#define LOG_MODULE_NAME app_lwm2m_accel
```
I fixed these across the whole sample.

mike-scott commented on samples/nrf9160/lwm2m_client/src/main.c:
```diff
@@ -0,0 +1,277 @@
+
+	len = strlen(buf);
+	for (i = 0; i < len; i++) {
+		if (imei_buf[i] >= 32 && imei_buf[i] <= 126) {
```
Nice catch.  I made this function more generic, but left the imei_buf references.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=1006) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=1006) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NordicPlayground/fw-nrfconnect-nrf?pullRequest=1006) it.</sub>

mike-scott commented:
@rlubos @mbolivar for review purposes here is the diff from the last version of the PR:
https://pastebin.com/raw/LN761Bm3

mike-scott commented:
One key difference being I removed the default mgmt.foundries.io Leshan server IP from the sample.  I generate a compile error if the user doesn't set that in prj.conf.

We can discuss if there's a better user-friendly way of setting that up.

mike-scott commented:
Updated with rough draft for documentation
diff: https://pastebin.com/raw/cEmqaUSW

rlubos commented on samples/nrf9160/lwm2m_client/src/main.c:
```diff
@@ -0,0 +1,277 @@
+#error "Missing CONFIG_LTE_LINK_CONTROL"
+#endif
+
+BUILD_ASSERT_MSG(sizeof(CONFIG_APP_LWM2M_SERVER) > 1,
```
Won't this check prevent the CI from building the sample successfully?

mike-scott replied:
Added a fix to pass CI (hopefully)

mike-scott commented:
New version posted.  Here's the diff: https://pastebin.com/raw/NpD9eHuX

Changes:
- Added `prj-ci.conf` which sets `CONFIG_APP_LWM2M_SERVER="passci"`
  I couldn't really come up with a better way of getting CI to pass while making sure the user changes this value.  I tried quite a few different methods but most were breaking due to the multi-image build style (IE: if I did this as an overlay the SPM build would complain that the overlay wasn't present.)
- Added the extra arg in sample.yml to pass CI: `extra_args: CONF_FILE="prj-ci.conf"`
- Fixed reference issue in accelerometer code
- Removed left over test printk from button code
- Removed currently unused function `double_to_float32_value()` from location code to pass CI build.

mike-scott commented:
updated west.yml to test this with PR: https://github.com/NordicPlayground/fw-nrfconnect-zephyr/pull/181

mike-scott commented:
will re-update to the mainline Zephyr SHA once the Zephyr PR is merged

mbolivar commented:
@rlubos can you take a look and merge if the docs issues in CI are resolved?

The error is internal to our system somehow (`requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://intranet.nordicsemi.no/TCM/ctf/wheel/`), not something related to this change.

cc: @carlescufi @thst-nordic

rlubos commented on samples/nrf9160/lwm2m_client/prj-ci.conf:
```diff
@@ -0,0 +1,112 @@
+CONFIG_NET_TX_STACK_SIZE=256
+
+# Set LwM2M Server IP address here
+CONFIG_APP_LWM2M_SERVER="passci"
```
How about `overlay-ci.conf` contatining:
```
# Set LwM2M Server IP address here
CONFIG_APP_LWM2M_SERVER="passci"
```
only? There's a risk that `prj.conf` && `prj-ci.conf` will diverse over time, and the CI will not build the actual sample configuration. You can apply the overlay like this:
`CONF_FILE="prj.conf overlay-ci.conf"`
Should also be possible with `OVERLAY_CONFGI=overlay-ci.conf`, though it had issues with multiimage build (fix was merged today into fw-nrfconnect-zephyr).

mike-scott replied:
You are correct, I was also nervous about separate conf file.  I will try overlay fix again today and send an updated patch (including zephyr SHA change).

rlubos commented:
The Zephyr dependency got merged, here's the SHA you need: https://github.com/NordicPlayground/fw-nrfconnect-zephyr/commit/7ff06506a50a70b32b69c4c9a669efb7ed232a48

thst-nordic commented:
CI/Jenkins/DOC has been disabled temporarily for an unrelated problem.   you can merge without it.

mike-scott commented:
@rlubos would it be acceptable to use the SHA for @SebastianBoe IMAGE fixes? 
https://github.com/NordicPlayground/fw-nrfconnect-zephyr/commit/9c29db6ede23b9a5de7b5cd74b5f659d83fca3f2

That way I can use `OVERLAY_CONFGI=overlay-ci.conf`?

mbolivar commented:
> @rlubos would it be acceptable to use the SHA for @SebastianBoe IMAGE fixes?
> NordicPlayground/fw-nrfconnect-zephyr@9c29db6

@mike-scott Yes, always use the latest master's SHA (9c29db6ede at time of writing as you say) if additional patches go in after your merge. Otherwise it removes necessary patches from history.

mike-scott commented:
Overlay now works correctly:
```
$ sanitycheck --platform nrf9160_pca10090ns --inline-logs --outdir /out/build --build-only --testcase-root=../nrf/samples/nrf9160/lwm2m_client/ --tag=ci_build
JOBS: 16
Cleaning output directory /out/build
Building testcase defconfigs...
Filtering test cases...
1 tests selected, 183 tests discarded due to filters
total complete:    1/   1  100%  failed:    0
1 of 1 tests passed with 0 warnings in 16 seconds
```

New patch posted.  Here's the diff: https://pastebin.com/raw/hXcrKgKi

mbolivar commented:
@mike-scott I just force-pushed into zephyr master to fix up the sauce tags on the latest commits, so I took the liberty of updating your branch as well. Small diff to the zephyr revision in west.yml:

```
-      revision: 9c29db6ede23b9a5de7b5cd74b5f659d83fca3f2
+      revision: 56f6a33a71c46ec5b2e1beb6076edf9521014ad3
```

Once CI passes (except the expected docs failure we are going to ignore), I will merge.

mike-scott commented:
@mbolivar sounds good
